### PR TITLE
[mojo-stdlib] Add `divmod(Int, Int)` to builtins

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
The documentation for python's divmod: 
* https://docs.python.org/3/library/functions.html#divmod
* https://docs.python.org/3/reference/datamodel.html#object.__divmod__

Later on, traits will allow us to generalize divmod to pretty much anything that has a `__divmod__` method, for example `datetime.timedelta`.

We also add tests here as the function `_divmod` of `Int` was previously untested. I verified that the output correspond to what the python interpreter returns when calling `divmod()`.

I don't know where to declare that a function doesn't need to be imported (available globally). Where can I find this? Is this declared in the compiler? [Divmod does not need to be imported in python]( https://docs.python.org/3/library/functions.html#built-in-functions)

### CODE REVIEW

**Comment by soraros** [comment]
I'm against adding this as currently there is no way to type `divmod` properly (similar to how we don't have `min`/`max`). You can call `__divmod__` directly on `datetime` as a workaround.

**Comment by gabrieldemarmiesse** [comment]
I respectfully disagree. It's a big quality of life improvement that has nearly no cost on the standard library side as the code added is tested, and the final implementation with traits will look very much like the current one, so very little modifications will be needed.

**Comment by soraros** [comment]
> we already have min and max, it only works with Int and SIMD at the moment

Exactly: not as "builtin" functions like in Python. I think adding a `divmod` function to `int` or `math` would be more fitting. Whether we should raise an error or trap (as with overflows) is another matter up for debate, though.

**Comment by ConnorGray** [comment]
Just a quick note, not weighing in more generally yet, but Mojo does have a [`math.divmod` function](https://docs.modular.com/mojo/stdlib/math/math#divmod), it's just not in the "builtin" / prelude.

**Comment by gabrieldemarmiesse** [comment]
Ho, you are totally right! I took the habit of using "search in files" rather than going in the documentation. Since the math module isn't open source, I didn't see divmod. I believe this PR is not necessary then. Good to know the work is already done :)

**Comment by ConnorGray** [comment]
No worries Gabriel, totally understand :)

Before I saw you closed this, I'd been planning to ask that we hold-off on any improvements in this area until we're able to marshal the math module into the open-source modules. Once that happens, changes like this will get easier for us to coordinate :) I'm not able to provide a firm estimate on when that might happen, but it is something we're planning for.

**Comment by gabrieldemarmiesse** [comment]
I Re-opened this PR to fix https://github.com/modularml/mojo/issues/2335 . Note that the implementation currently works for integers but not for other types yet. The trait is non-trivial and can be added in another PR. Notably, in Python, it's possible (and it makes sense) to divmod things that aren't abstract numbers, like time or currency, and in this case, the signature is more complexe. E.g. `divmod(some_euros, other_euros) = (some_int, euros_left)` and `divmod(some_euros, some_int) = (some_euros, euros_left)`

**Review by abduld — APPROVED** [review]
nice

File: stdlib/src/builtin/divmod.mojo [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ===----------------------------------------------------------------------=== #
+
+
+fn divmod(a: Int, b: Int) -> Tuple[Int, Int]:
```
[the remark this answers is not in the document]
Comment (laszlokindrat) [reply stdlib/src/builtin/divmod.mojo:15]: Actually, if you do this, please put it in a subsequent patch. And if you are around there, could you maybe add implementations for `{Float,Int}Literal` and potentially `SIMD`?

[the remark this answers is not in the document]
Comment (gabrieldemarmiesse) [reply stdlib/src/builtin/divmod.mojo]: It seems adding the trait makes us hit a thorny issue:

```
Included from /Users/runner/work/mojo/mojo/stdlib/src/builtin/__init__.mojo:1:
/Users/runner/work/mojo/mojo/stdlib/src/builtin/divmod.mojo:16:53: error: 'Tuple' parameter #0 has 'CollectionElement' type, but value has type 'DivModable'
    fn __divmod__(self: Self, other: Self) -> Tuple[Self, Self]:
```
If that's ok for you, can we merge this PR without the trait and we'll see in another PR how to add the trait and support for other types?

**Comment by gabrieldemarmiesse** [comment]
@laszlokindrat adding the trait was causing obscure issues. I reverted the change to avoid stalling this PR for too long. We can merge this and always add the trait in another PR. I can do a github issue to track it after this PR is merged.
For your information, here is the error message I got when adding the trait (with CollectionElement being a requirement):
```
/projects/open_source/mojo/stdlib/src/builtin/divmod.mojo:16:53: error: no '__copyinit__' candidates have type 'fn(inout self: T, existing: T, /) -> None'
    fn __divmod__(self: Self, other: Self) -> Tuple[Self, Self]:
                                                    ^~~~
```

**Comment by laszlokindrat** [comment]
@gabrieldemarmiesse Thanks again for the patch! I merged it without the trait. Do you think you could provide a small repro and file a ticket for the problem?

**Comment by gabrieldemarmiesse** [comment]
Sure, I'll open an issue about the trait and one issue about the bug to have a reprocible snippet

**Comment by gabrieldemarmiesse** [comment]
@laszlokindrat I made two follow-up issues to track the things discussed in this PR:
* #2597 
* #2598

**Pull request closed without merging**